dockstore-webservice/src/main/java/io/dockstore/webservice/resources/WorkflowResource.java

```diff
@@ -906,6 +950,355 @@ private void updateInfo(Workflow oldWorkflow, Workflow newWorkflow) {
+                        + " outputstream for DOI zip file for upload to Zenodo."
+                        + " Error is " + fne.getMessage(), fne);
+                throw new CustomWebApplicationException("Could not create file " + fileName
+                        + " outputstream for DOI zip file for upload to Zenodo."
+                        + " Error is " + fne.getMessage(), HttpStatus.SC_INTERNAL_SERVER_ERROR);
+            }
+            writeStreamAsZip(sourceFiles, outputStream, path);
+            java.nio.file.Path zipPath = Paths.get(fileName);
+            File zipFile = zipPath.toFile();
+
+            try {
+                filesApi.createFile(depositionID, zipFile, fileName);
+            } catch (ApiException e) {
+                LOG.error("Could not create files for new version on Zenodo. Error is " + e.getMessage(), e);
+                throw new CustomWebApplicationException("Could not create files for new version on Zenodo."
+                        + " Error is " + e.getMessage(), HttpStatus.SC_BAD_REQUEST);
+            }
+        }
+    }
+
+
+    /**
+     * Check if a Zenodo DOI already exists for the workflow version
+     * @param workflowVersion workflow version
+     */
+    private void checkForExistingDOIForWorkflowVersion(WorkflowVersion workflowVersion) {
+        String workflowVersionDoiURL = workflowVersion.getDoiURL();
+        if (workflowVersionDoiURL != null && !workflowVersionDoiURL.isEmpty()) {
+            LOG.error("Workflow version " + workflowVersion.getName() + " already has DOI " + workflowVersionDoiURL
+                    + ". Dockstore can only create one DOI per version.");
+            throw new CustomWebApplicationException("Workflow version " + workflowVersion.getName() + " already has DOI "
+                    + workflowVersionDoiURL + ". Dockstore can only create one DOI per version.", HttpStatus.SC_METHOD_NOT_ALLOWED);
+        }
+    }
+
+    /**
+     * Get an existing Zenodo DOI for the workflow if one exists
+     * otherwise return null
+     * @param workflow workflow
+     */
+    private String getAnExistingDOIForWorkflow(Workflow workflow) {
+        // Find out if this workflow already has at least one
+        // version that has been assigned a DOI
+        // If a version DOI exists, we will create another version DOI
+        // instead of creating a new workflow concept DOI and version DOI
+        // Get the ID of one of the workflow version DOIs
+        // because Zenodo requires that we use it to create the next
+        // workflow version DOI
+        String latestWorkflowVersionDOIURL = null;
+        Set<WorkflowVersion> setOfWorkflowVersions = workflow.getWorkflowVersions();
+        Iterator iter = setOfWorkflowVersions.iterator();
+        while (iter.hasNext()) {
+            WorkflowVersion myWorkflowVersion = ((WorkflowVersion)iter.next());
+            latestWorkflowVersionDOIURL = myWorkflowVersion.getDoiURL();
```

Comment:
If the last version in the set has an empty DOI url, you will return an empty string instead of null. Since your next line checks for empty, it sounds like you don't want to do that.

In fixing this, you might just want to use a stream and find-- I had to read the existing loop a few times to understand it, whereas the stream/find is easier to read and understand (and likely this issue wouldn't have occurred).

Update: I see below on line 1245 that you check if the return value is null or empty. It's still confusing though, because your JavaDoc says it returns the DOI or null, but then you return DOI, "", or null. But do a test here on the next line that makes sure it's not "".

Suggest returning either null or a DOI (or even better, `Optional<String>`), but not return nor need to check for 2 different ways of saying the DOI is not set.

Reply:
Next sprint https://ucsc-cgl.atlassian.net/browse/SEAB-350